rakelib/service_tags.rake

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+namespace :service_tags do
+  def changed_files
+    ENV['CHANGED_FILES'] || []
+  end
+
+  def controller_info_from_route(route)
+    return unless route.defaults[:controller]
+
+    controller_name = "#{route.defaults[:controller].camelize}Controller"
+    return unless Object.const_defined?(controller_name)
+
+    controller_class = controller_name.constantize
+    exclusive_methods = controller_class.instance_methods(false)
+    return if exclusive_methods.empty?
+
+    method_name = exclusive_methods.first
+    file_path = controller_class.instance_method(method_name).source_location.first
+    relative_path = Pathname.new(file_path).relative_path_from(Rails.root).to_s
+
+    {
+      name: controller_name,
+      path: relative_path
+    }
+  end
+
+  def controllers_from_routes(routes)
+    routes.map { |route| controller_info_from_route(route) }.compact.uniq { |info| info[:name] }
+  end
+
+  def valid_service_tag?(klass)
+    klass.ancestors.any? do |ancestor|
+      ancestor.included_modules.include?(Traceable) &&
+        ancestor.respond_to?(:trace_service_tag) &&
+        ancestor.try(:trace_service_tag).present?
+    end
+  end
+
+  def find_invalid_controllers(controllers)
+    errors = []
+    warnings = []
+
+    controllers.each do |controller|
+      excluded_prefixes = %w[ActionMailbox:: ActiveStorage:: ApplicationController OkComputer:: Rails::].freeze
```

Comment:
We have some number of controllers that are themselves called `ApplicationController`: 

```modules/simple_forms_api/app/controllers/simple_forms_api/v1/application_controller.rb:    class ApplicationController < ::ApplicationController
modules/my_health/app/controllers/my_health/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/test_user_dashboard/app/controllers/test_user_dashboard/application_controller.rb:  class ApplicationController < ActionController::API
modules/facilities_api/app/controllers/facilities_api/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/income_limits/app/controllers/income_limits/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/ask_va_api/app/controllers/ask_va_api/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/va_forms/app/controllers/va_forms/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/covid_research/app/controllers/covid_research/application_controller.rb:  class ApplicationController < ActionController::API
modules/covid_vaccine/app/controllers/covid_vaccine/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/vba_documents/app/controllers/vba_documents/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/veteran_verification/app/controllers/veteran_verification/application_controller.rb:  class ApplicationController < ::OpenidApplicationController
modules/claims_api/app/controllers/claims_api/v2/application_controller.rb:    class ApplicationController < ::ApplicationController
modules/claims_api/app/controllers/claims_api/v1/application_controller.rb:    class ApplicationController < ::ApplicationController
modules/appeals_api/app/controllers/appeals_api/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/check_in/app/controllers/check_in/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/apps_api/app/controllers/apps_api/application_controller.rb:  class ApplicationController < ::OpenidApplicationController
modules/debts_api/app/controllers/debts_api/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/mobile/app/controllers/mobile/application_controller.rb:  class ApplicationController < SignIn::ApplicationController
modules/veteran_confirmation/app/controllers/veteran_confirmation/application_controller.rb:  class ApplicationController < ::ApplicationController
modules/mocked_authentication/app/controllers/mocked_authentication/application_controller.rb:  class ApplicationController < ActionController::Base
app/controllers/application_controller.rb:class ApplicationController < ActionController::API
app/controllers/sign_in/application_controller.rb:  class ApplicationController < ActionController::API
```

Will this exclude them from the check? Should it?

Reply:
@acrollet The rake task only flags controllers that have methods wired up to routes. The only controller from that list that has a route configured is the Rails app's base application controller which has an OPTIONS method call to application#cors_preflight. Since that's the base controller for 146 other controllers and a service tag there would be inherited (e.g. the mobile module), and be compliant (but not 😄) so we should exclude the application controller so its descendants are flagged.

It does mean if a team wanted to break rails conventions and connect a route to an application controller they could through a named route, but not with resource based routes as they look for a {resource}s_controller.rb

I'd need to investigate it further but I don't think cors_preflight has to go in the main application controller so we could move it into its own.

Reply:
That makes sense, thanks for spelling it out for this rails n00b 😁